modules/repository-azure/src/main/java/org/elasticsearch/repositories/azure/AzureClientProvider.java

```diff
@@ -259,38 +264,136 @@ protected void doStop() {
     protected void doClose() {}
 
-    private static final class SuccessfulRequestTracker implements HttpPipelinePolicy {
-        private static final Logger logger = LogManager.getLogger(SuccessfulRequestTracker.class);
+    static class RequestMetrics {
+        private volatile long timeToResponseInMillis = -1;
+        private volatile int requestCount;
+        private volatile int errorCount;
+        private volatile int throttleCount;
+        private volatile int statusCode;
+
+        int getRequestCount() {
+            return requestCount;
+        }
+
+        int getErrorCount() {
+            return errorCount;
+        }
+
+        int getStatusCode() {
+            return statusCode;
+        }
+
+        int getThrottleCount() {
+            return throttleCount;
+        }
+
+        /**
+         * Time to response, or -1 if response was never received
+         */
+        long getTimeToResponseInMillis() {
+            return timeToResponseInMillis;
+        }
+
+        @Override
+        public String toString() {
+            return "RequestMetrics{"
+                + "timeToResponseInMillis="
+                + timeToResponseInMillis
+                + ", requestCount="
+                + requestCount
+                + ", errorCount="
+                + errorCount
+                + ", throttleCount="
+                + throttleCount
+                + ", statusCode="
+                + statusCode
+                + '}';
+        }
+    }
+
+    private enum RetryMetricsTracker implements HttpPipelinePolicy {
+        INSTANCE;
+
+        @Override
+        public Mono<HttpResponse> process(HttpPipelineCallContext context, HttpPipelineNextPolicy next) {
+            RequestMetrics metrics = (RequestMetrics) context.getData(RequestMetricsTracker.ES_REQUEST_METRICS_CONTEXT_KEY)
+                .orElseGet(() -> {
+                    String errorMessage = "No metrics object associated with request " + context.getHttpRequest();
+                    assert false : errorMessage;
+                    throw new IllegalStateException(errorMessage);
+                });
+            metrics.requestCount++;
+            return next.process().doOnError(throwable -> {
+                logger.debug("Detected error in RetryMetricsTracker", throwable);
+                metrics.errorCount++;
+            }).doOnSuccess(response -> {
+                if (isErrorCode(response.getStatusCode())) {
+                    metrics.errorCount++;
+                    if (response.getStatusCode() == RestStatus.TOO_MANY_REQUESTS.getStatus()) {
+                        metrics.throttleCount++;
+                    }
+                }
+            });
+        }
+
+        @Override
+        public HttpPipelinePosition getPipelinePosition() {
+            return HttpPipelinePosition.PER_RETRY;
+        }
+
+        private static boolean isErrorCode(int statusCode) {
+            return statusCode <= 199 || statusCode > 299;
+        }
```

Comment:
This will record an "error" for e.g. a 404 that we get when checking if a blob exists. Technically that's probably not an "error", but it seems consistent with the S3 implementation.